pkg/global/apply_configuration.go

```diff
@@ -43,43 +47,94 @@ import (
 // successfully.
-type LifecycleState struct {
+type DiagnosticsServer struct {
 	config                          *pb.DiagnosticsHTTPServerConfiguration
 	activeSpansReportingHTTPHandler *bb_otel.ActiveSpansReportingHTTPHandler
+	state                           int32
+	server                          *http.Server
 }
 
-// MarkReadyAndWait can be called to report that the program has started
-// successfully. The application should now be reported as being healthy
-// and ready, and receive incoming requests if applicable.
-func (ls *LifecycleState) MarkReadyAndWait() {
+// Serve can be called to report that the program has started successfully.
+// The application should now be reported as being healthy and ready, according
+// to isReady, and receive incoming requests if applicable.
+func (ds *DiagnosticsServer) Serve(terminationContext context.Context) error {
 	// Start a diagnostics web server that exposes Prometheus
 	// metrics and provides a health check endpoint.
-	if ls.config == nil {
-		select {}
-	} else {
+	if ds.config != nil {
 		router := mux.NewRouter()
 		router.HandleFunc("/-/healthy", func(http.ResponseWriter, *http.Request) {})
-		if ls.config.EnablePrometheus {
+		router.HandleFunc("/-/ready", func(w http.ResponseWriter, _ *http.Request) {
+			if ds.state == stateServing {
+				w.WriteHeader(http.StatusOK)
+			} else {
+				http.Error(w, http.StatusText(http.StatusServiceUnavailable), http.StatusServiceUnavailable)
+			}
+		})
+		if ds.config.EnablePrometheus {
 			router.Handle("/metrics", promhttp.Handler())
 		}
-		if ls.config.EnablePprof {
+		if ds.config.EnablePprof {
 			router.PathPrefix("/debug/pprof/").Handler(http.DefaultServeMux)
 		}
-		if httpHandler := ls.activeSpansReportingHTTPHandler; httpHandler != nil {
+		if httpHandler := ds.activeSpansReportingHTTPHandler; httpHandler != nil {
 			router.Handle("/active_spans", httpHandler)
 		}
 
-		log.Fatal(http.ListenAndServe(ls.config.ListenAddress, router))
+		ds.server = &http.Server{
+			Addr:    ds.config.ListenAddress,
+			Handler: router,
+		}
+		go func() {
+			<-terminationContext.Done()
+			ds.SetNotServing()
+			ds.server.Shutdown(terminationContext)
+		}()
+		if err := ds.server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			return err
+		}
+	} else {
+		<-terminationContext.Done()
 	}
+	return nil
+}
+
+// SetReady updates the health probe to report healthy and ready.
+func (ds *DiagnosticsServer) SetReady() {
+	ds.state = stateServing
+}
+
+// SetNotServing updates the health probe to report healthy but not ready.
+func (ds *DiagnosticsServer) SetNotServing() {
+	ds.state = stateNotServing
+}
+
+// ServeDiagnostics is a wrapper that calls DiagnosticsServer.Serve inside
+// a goroutine, managed by the provided errgroup.Group, and returns
+// immediately.
+func ServeDiagnostics(terminationContext context.Context, terminationGroup *errgroup.Group, diagnosticsServer *DiagnosticsServer) {
+	terminationGroup.Go(func() error {
```

Comment:
To me it's not completely clear why we need to use `terminationGroup` here.

- `terminationGroup` is what we need to register tasks that need to extend the lifetime of the process in case of ^C.
- `terminationContext` is what we need to detect that termination has started.

I think we only need to store the `terminationContext`, and let `/-/healthy` do this:
```go
switch {
case <-ctx.Done():
    // Return a HTTP 500, as we're shutting down.
default:
    // Return a HTTP 200.
}
```

Reply:
If either the diagnostics or gRPC servers fail, then the `terminationGroup` will be cancelled and the data will be synced. Using a standard goroutine together with `log.Fatal` there might be data loss. I don't know why they would fail, but for me that sounds reasonable. What do you think?